desktop-widgets/mainwindow.cpp

```diff
@@ -89,23 +89,40 @@ extern "C" int updateProgress(const char *text)
 
 MainWindow *MainWindow::m_Instance = NULL;
 
-extern "C" void showErrorFromC()
+extern "C" void showErrorFromC(char *buf)
 {
-	// Show errors only if we are running in the GUI thread.
-	// If we're not in the GUI thread, let errors accumulate.
-	if (QThread::currentThread() != QCoreApplication::instance()->thread())
-		return;
+	QString error(buf);
+	free(buf);
 
 	MainWindow *mainwindow = MainWindow::instance();
-	if (mainwindow)
-		mainwindow->showErrors();
+	// This test is redundant, because the callback is registered after creation of the instance
+	if (mainwindow) {
+		mainwindow->registerError(error);
+		// If we're in the GUI thread show the error immediately
+		if (QThread::currentThread() == QCoreApplication::instance()->thread())
+			mainwindow->showErrors();
+	}
+}
+
+void MainWindow::registerError(const QString &s)
+{
+	QMutexLocker l(&errorMutex);
+	errors << s;
+}
+
+QStringList MainWindow::consumeErrors()
+{
+	QStringList ret;
+	QMutexLocker l(&errorMutex);
+	std::swap(errors, ret);
+	return ret;
 }
 
 void MainWindow::showErrors()
 {
-	const char *error = get_error_string();
-	if (!empty_string(error))
-		getNotificationWidget()->showNotification(error, KMessageWidget::Error);
+	QStringList errors = consumeErrors();
+	while (!errors.isEmpty())
+		getNotificationWidget()->showNotification(errors.takeFirst(), KMessageWidget::Error);
```

Comment:
it will just go through the loop and show the last error. this can be seen in `NotificationWidget', so i think there is no point in accumulating them with a mutex. i will make another proposal.